### PR TITLE
create playlist

### PR DESCRIPTION
**Pull request by bobobob22**

### CODE REVIEW

File: packages/webapp/src/store/router/router.saga.ts [hunk 1]
```diff
@@ -5,7 +5,7 @@ import { Router } from '../consts/actionTypes'
 import { RouterActions } from './router.reducer'
 
 export function* push(history: History, action: RouterActions['historyPush']): Generator {
-  yield history.push(action.payload)
+  // yield history.push(action.payload)
```
Comment (kowalski) [inline packages/webapp/src/store/router/router.saga.ts:8]: intentional change?

File: packages/webapp/src/store/uploadSongs/uploadSongs.reducer.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  loadFile: createAction<File>(UploadSongs.LOAD_FILE),
+  saveSongProgress: createAction<number>(UploadSongs.SAVE_SONG_PROGRESS),
+  updateSongProgress: createAction<ISongProgress>(UploadSongs.UPDATE_SONG_PROGRESS),
+}
+
+export const uploadSongsReducer = createReducer(initialState as UploadSongsStore, (builder) => {
+  builder
+    .addCase(uploadSongsActions.addFile, (state, action: PayloadAction<IUploadingFile>) => {
+      const prevSongs = [...state.songs]
+      state.songs = [...prevSongs, action.payload]
```
Comment (kowalski) [inline packages/webapp/src/store/uploadSongs/uploadSongs.reducer.ts:30]: you don't need this unpacking
reduxjs/toolkit uses internally `immer` that ensures this

its sufficient to do:
```
state.songs.push(action.payload)
```

Comment (kowalski) [reply packages/webapp/src/store/uploadSongs/uploadSongs.reducer.ts:30]: same applies to other actions in this reducer

File: packages/webapp/src/store/uploadSongs/uploadSongs.saga.ts [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+      duration = secondsToMinutes(Number((event.currentTarget as HTMLAudioElement).duration))
+      resolve(duration)
+
+      return duration
+    })
+  })
+}
+
+async function getSongData(currentFile: File): Promise<TagType> {
+  const promise = new Promise<TagType>((resolve, reject) => {
```
Comment (kowalski) [inline packages/webapp/src/store/uploadSongs/uploadSongs.saga.ts:30]: actually you could just be doing 
```
return new Promise(....)
```

without this
```
 const data = await promise

  return data
```
this two lines have literally no effect

File: packages/webapp/src/store/uploadSongs/uploadSongs.saga.ts [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+  while (true) {
+    const { transferredData = 0, error, success } = yield take(channel)
+    if (error) {
+      console.log(error)
+
+      return
+    }
+    if (success) {
+      console.log('success', success)
+
```
Comment (kowalski) [inline packages/webapp/src/store/uploadSongs/uploadSongs.saga.ts:91]: do we want to return here?
perhaps we should put progress to 100 first?

File: packages/webapp/src/store/uploadSongs/uploadSongs.selector.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+  songs,
+  songsProgress,
+}
+
+export const uploadingSongsWithProgress = createSelector(
+  [uploadSongsSelectors.songs, uploadSongsSelectors.songsProgress],
+  (songsArray: IUploadingFile[], songsProgressArray: ISongProgress[]) => {
+    const songsWithCurrentProgress = songsArray.map((song) => ({
+      ...song,
+      ...songsProgressArray.find((progress) => progress.id === song.id),
```
Comment (kowalski) [inline packages/webapp/src/store/uploadSongs/uploadSongs.selector.ts:23]: wouldn't it be easier if `songsProgressArray` is a `Map<number, ISongProgress>` ?

Comment (kowalski) [reply packages/webapp/src/store/uploadSongs/uploadSongs.selector.ts:23]: (and therefore is called `songsProgressMap`

**Review by kowalski — CHANGES_REQUESTED** [review]

File: packages/webapp/src/store/uploadSongs/uploadSongs.reducer.ts (hunk 2 above)
Comment (bobobob22) [reply packages/webapp/src/store/uploadSongs/uploadSongs.reducer.ts:30]: ok

File: packages/webapp/src/store/uploadSongs/uploadSongs.saga.ts (hunk 3 above)
Comment (bobobob22) [reply packages/webapp/src/store/uploadSongs/uploadSongs.saga.ts:29]: ok

File: packages/webapp/src/store/uploadSongs/uploadSongs.saga.ts (hunk 4 above)
Comment (bobobob22) [reply packages/webapp/src/store/uploadSongs/uploadSongs.saga.ts:93]: ok

File: packages/webapp/src/store/uploadSongs/uploadSongs.selector.ts (hunk 5 above)
Comment (bobobob22) [reply packages/webapp/src/store/uploadSongs/uploadSongs.selector.ts]: ok

File: packages/webapp/src/store/router/router.saga.ts (hunk 1 above)
Comment (bobobob22) [reply packages/webapp/src/store/router/router.saga.ts]: only for testing, i reversed it

**Pull request merged**